File: frontend/src/lib/resumeDiff.ts
/**
 * What one application's resume changed relative to the master it came from.
 *
 * The server already sends *why* it changed (`changes`, written by the
 * derivation). This computes *what* changed, item by item, so the difference
 * can be shown against the actual document rather than described in prose —
 * and so an edit the user made by hand shows up too, which the server's change
 * list, written at derivation time, would not mention.
 *
 * Both documents come down in the same payload, so this is a pure function over
 * data already in hand: no request, no server-side diff endpoint.
 */

import type { ResumeDocument, ResumeExperience } from "@/types/api";

/** Items whose position moved forward, in their new order. */
function promoted(base: string[], next: string[]): string[] {
  return next.filter((item, index) => {
    const before = base.indexOf(item);
    return before > index;
  });
}

function sameOrder(base: string[], next: string[]): boolean {
  return base.length === next.length && base.every((item, index) => item === next[index]);
}

export interface ExperienceDiff {
  key: string;
  company: string;
  role: string;
  /** 1-based, so it reads as "3rd → 1st" without arithmetic in the view. */
  basePosition: number;
  position: number;
  /** Positive when this experience moved up the resume. */
  movedUpBy: number;
  summaryChanged: boolean;
  baseSummary: string | null;
  summary: string | null;
  /** The posting's terms this experience is being used to answer. */
  focus: string[];
  highlightsReordered: boolean;
  /** Achievements folded away because nothing here matched the posting. */
  highlightsHidden: number;
  /** Achievement text the master does not have — only a hand edit can do this. */
  highlightsAdded: string[];
  technologiesPromoted: string[];
  technologiesAdded: string[];
  technologiesRemoved: string[];
  projectsReordered: boolean;
  changed: boolean;
}

export interface ResumeDiff {
  summaryChanged: boolean;
  baseSummary: string | null;
  summary: string | null;
  experiencesReordered: boolean;
  skillsPromoted: string[];
  technologiesPromoted: string[];
  projectsReordered: boolean;
  experiences: ExperienceDiff[];
  /** Experiences the master has that this version dropped entirely. */
  droppedExperiences: string[];
  changedCount: number;
  hasChanges: boolean;
}

function experienceDiff(
  base: ResumeExperience,
  next: ResumeExperience,
  basePosition: number,
  position: number,
): ExperienceDiff {
  const baseHighlights = base.highlights.map((item) => item.text);
  const nextHighlights = next.highlights.map((item) => item.text);
  const kept = nextHighlights.filter((text) => baseHighlights.includes(text));

  const technologiesAdded = next.technologies.filter(
    (item) => !base.technologies.includes(item),
  );
  const technologiesRemoved = base.technologies.filter(
    (item) => !next.technologies.includes(item),
  );

  const diff: ExperienceDiff = {
    key: next.key,
    company: next.company,
    role: next.role,
    basePosition,
    position,
    movedUpBy: basePosition - position,
    summaryChanged: (base.summary ?? "") !== (next.summary ?? ""),
    baseSummary: base.summary,
    summary: next.summary,
    focus: next.focus,
    // Reordering is only meaningful among the achievements that survived.
    highlightsReordered: !sameOrder(
      baseHighlights.filter((text) => kept.includes(text)),
      kept,
    ),
    highlightsHidden: baseHighlights.filter((text) => !nextHighlights.includes(text)).length,
    highlightsAdded: nextHighlights.filter((text) => !baseHighlights.includes(text)),
    technologiesPromoted: promoted(base.technologies, next.technologies),
    technologiesAdded,
    technologiesRemoved,
    projectsReordered: !sameOrder(
      base.projects.map((item) => item.name),
      next.projects.map((item) => item.name),
    ),
    changed: false,
  };

  diff.changed =
    diff.movedUpBy !== 0 ||
    diff.summaryChanged ||
    diff.highlightsReordered ||
    diff.highlightsHidden > 0 ||
    diff.highlightsAdded.length > 0 ||
    diff.technologiesPromoted.length > 0 ||
    technologiesAdded.length > 0 ||
    technologiesRemoved.length > 0 ||
    diff.projectsReordered;

  return diff;
}

/**
 * Compare an application's resume against the master snapshot it was built from.
 *
 * The snapshot, not the live master: this answers "how does this application
 * present me differently", and comparing against a profile the user edited
 * afterwards would blame the version for changes it never made. Staleness
 * against the live master is a separate flag the server sends.
 */
export function diffResume(base: ResumeDocument, document: ResumeDocument): ResumeDiff {
  const basePositions = new Map(base.experiences.map((item, index) => [item.key, index]));

  const experiences = document.experiences.map((experience, index) => {
    const basePosition = basePositions.get(experience.key);
    const source =
      basePosition === undefined ? experience : base.experiences[basePosition] ?? experience;
    return experienceDiff(
      source,
      experience,
      (basePosition ?? index) + 1,
      index + 1,
    );
  });

  const documentKeys = new Set(document.experiences.map((item) => item.key));
  const diff: ResumeDiff = {
    summaryChanged: (base.summary ?? "") !== (document.summary ?? ""),
    baseSummary: base.summary,
    summary: document.summary,
    experiencesReordered: !sameOrder(
      base.experiences.map((item) => item.key),
      document.experiences.map((item) => item.key),
    ),
    skillsPromoted: promoted(base.skills, document.skills),
    technologiesPromoted: promoted(base.technologies, document.technologies),
    projectsReordered: !sameOrder(
      base.projects.map((item) => item.name),
      document.projects.map((item) => item.name),
    ),
    experiences,
    droppedExperiences: base.experiences
      .filter((item) => !documentKeys.has(item.key))
      .map((item) => `${item.role} — ${item.company}`),
    changedCount: 0,
    hasChanges: false,
  };

  diff.changedCount =
    experiences.filter((item) => item.changed).length +
    (diff.summaryChanged ? 1 : 0) +
    (diff.skillsPromoted.length > 0 ? 1 : 0) +
    (diff.technologiesPromoted.length > 0 ? 1 : 0) +
    (diff.projectsReordered ? 1 : 0);
  diff.hasChanges = diff.changedCount > 0 || diff.droppedExperiences.length > 0;

  return diff;
}

File: backend/app/domain/resume_diff.py
"""Your resume, and the one this vacancy gets — side by side.

The derivation already reports *that* it reorganised something
(`app.domain.resume.ResumeChange`). What a reviewer actually needs before
approving is narrower and more checkable: which experience moved from third to
first, which technologies came forward, how many sentences were promoted — and,
above all, **how few** edits there were.

That last part is the product rule this module exists to make visible. An
adapted resume has to still read as the same document. So the comparison ends
in a budget the reviewer can check at a glance:

    3 alterações · 2 tecnologias destacadas · 1 trecho ajustado · 0 invenções

The fourth number is not an assumption. `invented` is computed by running the
invention guard over the adapted document against the master's own words, so a
zero there is a measurement rather than a promise — and a non-zero one is the
signal to stop and read.

Pure: two documents in, counts and moves out. No database, no model call.
"""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass, field

from app.domain.language import fold
from app.domain.technologies import flag_unsupported_skills

# Terms named in a single "moved forward" line. Past this the line stops being
# a summary of what changed and becomes the list itself.
MAX_NAMED_TERMS = 6


@dataclass(frozen=True, slots=True)
class SnapshotExperience:
    """One position as an application's stored copy presents it.

    Deliberately a plain shape rather than the ORM row or the response model:
    what is stored is JSON, and the comparison has to work on exactly that.
    """

    experience_id: int | None = None
    company: str = ""
    role: str = ""
    summary: str = ""
    responsibilities: tuple[str, ...] = ()
    technologies: tuple[str, ...] = ()
    results: tuple[str, ...] = ()
    matched_terms: tuple[str, ...] = ()
    promoted: int = 0


@dataclass(frozen=True, slots=True)
class Snapshot:
    """The document one application presents, as stored."""

    summary: str = ""
    skills: tuple[str, ...] = ()
    emphasized_technologies: tuple[str, ...] = ()
    experiences: tuple[SnapshotExperience, ...] = ()

    def searchable(self) -> str:
        """Every word this document renders, for the invention guard."""
        parts: list[str] = [self.summary, *self.skills, *self.emphasized_technologies]
        for experience in self.experiences:
            parts += [experience.company, experience.role, experience.summary]
            parts += experience.responsibilities
            parts += experience.technologies
            parts += experience.results
        return "\n".join(part for part in parts if part)


@dataclass(frozen=True, slots=True)
class ExperienceMove:
    """One position that changed place, and why.

    Positions are 1-based, because they are read as "3ª → 1ª" and nobody counts
    their own CV from zero.
    """

    experience_id: int | None
    company: str
    role: str
    from_position: int
    to_position: int
    promoted_bullets: int = 0
    matched_terms: tuple[str, ...] = ()

    @property
    def moved(self) -> bool:
        return self.from_position != self.to_position


@dataclass(frozen=True, slots=True)
class ResumeComparison:
    """What separates the master resume from this application's copy.

    `is_comparable` is false when the copy was derived from a different master
    than the one being compared against — the user edited their profile
    afterwards. The diff would then attribute their own edits to the
    adaptation, which is worse than showing nothing and saying why.
    """

    moves: tuple[ExperienceMove, ...] = ()
    highlighted_technologies: tuple[str, ...] = ()
    promoted_bullets: int = 0
    # Terms in the adapted document that the master's own words cannot back.
    # Expected empty, measured rather than assumed.
    invented: tuple[str, ...] = field(default_factory=tuple)
    is_comparable: bool = True

    @property
    def experiences_reordered(self) -> int:
        return sum(1 for move in self.moves if move.moved)

    @property
    def sections_adjusted(self) -> int:
        """Positions whose own sentences were reordered inside them."""
        return sum(1 for move in self.moves if move.promoted_bullets > 0)

    @property
    def changes_total(self) -> int:
        """The headline number: how many distinct edits this copy carries."""
        return (
            self.experiences_reordered
            + len(self.highlighted_technologies)
            + self.sections_adjusted
        )

    @property
    def is_clean(self) -> bool:
        """Nothing in the copy that the master cannot back."""
        return not self.invented


def _moved_forward(master_order: Sequence[str], adapted_order: Sequence[str]) -> tuple[str, ...]:
    """Terms the adapted document lists earlier than the master does.

    Only forward moves count. A term that slipped down did so because another
    one came up, and reporting both would double every highlight.
    """
    positions = {fold(term): index for index, term in enumerate(master_order)}
    forward: list[str] = []
    for index, term in enumerate(adapted_order):
        previous = positions.get(fold(term))
        if previous is not None and index < previous:
            forward.append(term)
    return tuple(forward[:MAX_NAMED_TERMS])


def compare(
    *,
    master_experience_ids: Sequence[int | None],
    master_skills: Sequence[str],
    master_text: str,
    snapshot: Snapshot,
    is_comparable: bool = True,
) -> ResumeComparison:
    """Diff one application's copy against the master it came from.

    `master_experience_ids` is the master's own display order; the copy's
    `experience_id` values are matched against it, which is why the derivation
    keeps that id for provenance even though it never dereferences it.

    A position the master no longer has is skipped rather than reported as a
    move: it was deleted from the profile after this copy was frozen, and the
    copy keeping it is the documented behaviour, not a change the adaptation
    made.
    """
    if not is_comparable:
        return ResumeComparison(is_comparable=False)

    order = {value: index for index, value in enumerate(master_experience_ids) if value is not None}

    moves: list[ExperienceMove] = []
    for index, experience in enumerate(snapshot.experiences):
        previous = order.get(experience.experience_id) if experience.experience_id else None
        if previous is None:
            continue
        moves.append(
            ExperienceMove(
                experience_id=experience.experience_id,
                company=experience.company,
                role=experience.role,
                from_position=previous + 1,
                to_position=index + 1,
                promoted_bullets=experience.promoted,
                matched_terms=experience.matched_terms[:MAX_NAMED_TERMS],
            )
        )

    highlighted = _moved_forward(master_skills, snapshot.skills)
    promoted = sum(experience.promoted for experience in snapshot.experiences)
    invented = tuple(flag_unsupported_skills(master_text, snapshot.searchable()))

    return ResumeComparison(
        moves=tuple(moves),
        highlighted_technologies=highlighted,
        promoted_bullets=promoted,
        invented=invented,
        is_comparable=True,
    )


__all__ = [
    "MAX_NAMED_TERMS",
    "ExperienceMove",
    "ResumeComparison",
    "Snapshot",
    "SnapshotExperience",
    "compare",
]
